staging/src/k8s.io/apimachinery/pkg/util/net/http.go

```diff
@@ -189,6 +194,46 @@ func GetClientIP(req *http.Request) net.IP {
 	return net.ParseIP(req.RemoteAddr)
 }
 
+// Prepares the X-Forwarded-For header for another forwarding hop by appending the previous senders
+// IP address to the X-Forwarded-For chain.
+func PrepareForwardedFor(req *http.Request) {
+	remoteIP := GetRemoteIP(req)
+
+	hdr := req.Header
+	// First check the X-Forwarded-For header for requests via proxy.
+	hdrForwardedFor := hdr.Get("X-Forwarded-For")
+	if hdrForwardedFor != "" {
+		// Append the remote IP if necessary.
+		parts := strings.Split(hdrForwardedFor, ",")
+		for i := len(parts) - 1; i >= 0; i-- {
+			// Find the last non-empty part.
+			if parts[i] == "" {
+				continue
+			}
+			if ip := net.ParseIP(strings.TrimSpace(parts[i])); ip.Equal(remoteIP) {
+				// Remote IP already appended; nothing to do.
+				return
+			}
+			parts = append(parts[:i+1], remoteIP.String())
+			hdr.Set("X-Forwarded-For", strings.Join(parts, ","))
+			return
+		}
+		// Empty header, handle as if the header was not set.
+		hdrForwardedFor = ""
+	}
+
+	// Try the X-Real-Ip header.
+	hdrRealIp := hdr.Get("X-Real-Ip")
```

Comment:
I don't believe any originating IP address can be trusted. That having been said X-Real-Ip is meant to be the originating IP. I don't believe its clear whether GetRemoteIP() is the originating IP, the last IP in the proxy chain or something in between.  This is exacerbated by systems line HAProxy and NGINX which are configurable to do either. X-Forwarded-For is nice because while its no more trust worthy it should have all but the proxy IP in its list (assuming that no one has improperly mucked with it). Of course it assumes that the last proxy IP should be the result of calling GetRemoteIP().